fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeCollection.java

```diff
@@ -0,0 +1,172 @@
+                    // we should check for each action, for those that's not in the existing entry
+                    // we should create a new entry or added to the matching one
+                    ActionSet remaining = oppositeEntry.actionSet.difference(actionSet);
+                    if (! remaining.isEmpty()) {
+                        addAction(privilegeEntryList, entry, remaining, object, isGrant);
+                    }
+                }
+            }
+        } // for object in objects
+    }
+
+    public void revoke(short type, ActionSet actionSet, List<PEntryObject> objects, boolean isGrant) {
+        if (!typeToPrivilegeEntryList.containsKey(type)) {
+            LOG.debug("revoke a non-existence type {}", type);
+            return;
+        }
+        List<PrivilegeEntry> privilegeEntryList = typeToPrivilegeEntryList.get(type);
+        for (PEntryObject object : objects) {
+            PrivilegeEntry entry = findEntry(privilegeEntryList, object, isGrant);
+            if (entry != null) {
+                removeAction(privilegeEntryList, entry, actionSet);
+            }
+            // some of the actions may not granted
+            entry = findEntry(privilegeEntryList, object, !isGrant);
+            if (entry != null) {
+                // 1. intend to revoke with grant option but already grant object without grant option
+                // 2. intend to revoke without grant option but already grant object with grant option
+                // either way, we should remove the action here
+                removeAction(privilegeEntryList, entry, actionSet);
+            }
+        }
+        if (privilegeEntryList.isEmpty()) {
+            typeToPrivilegeEntryList.remove(type);
+        }
+    }
+
+    public boolean check(short type, Action want, PEntryObject object) {
+        if (!typeToPrivilegeEntryList.containsKey(type)) {
+            return false;
+        }
+        List<PrivilegeEntry> privilegeEntryList = typeToPrivilegeEntryList.get(type);
+        for (PrivilegeEntry privilegeEntry : privilegeEntryList) {
+            if (privilegeEntry.object.equals(object) && privilegeEntry.actionSet.contains(want)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean checkAnyObject(short type, Action want) {
+        if (!typeToPrivilegeEntryList.containsKey(type)) {
+            return false;
+        }
+        List<PrivilegeEntry> privilegeEntryList = typeToPrivilegeEntryList.get(type);
+        for (PrivilegeEntry privilegeEntry : privilegeEntryList) {
+            if (privilegeEntry.actionSet.contains(want)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean hasType(short type) {
+        return typeToPrivilegeEntryList.containsKey(type);
+    }
+
+    public boolean allowGrant(short type, Action want, PEntryObject object) {
```

Comment:
Since these three functions `allowGrant`, `check`, `checkAnyObject` are highly similar, shall we merge them together?

Reply:
No, these are plugin interfaces meant to override by developers.